server/storage/mvcc/watchable_store.go

```diff
@@ -485,6 +487,29 @@ func (s *watchableStore) progress(w *watcher) {
 	}
 }
 
+func (s *watchableStore) progressAll(watchers map[WatchID]*watcher) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	// Any watcher unsynced?
+	for w, _ := range watchers {
+		if _, ok := s.synced.watchers[watchers[w]]; !ok {
+			return false
+		}
+	}
+
+	// If all watchers are synchronised, send out progress
+	// notification on first watcher. Note that all watchers
+	// should have the same underlying stream, and the progress
+	// notification will be broadcasted client-side if required
+	// (see dispatchEvent in client/v3/watch.go)
```

Comment:
This is not obvious when calling function. For me this function is `progressIfSynced(w *watcher, requiredSynced []*watcher)`. I would even go further and merge it with current `progress` function. In both cases we send progress when set of watchers are synced. Function `progress` just has `requiredSynced = []*watcher{w}`. 

Something like
```
func (s *watchableStore) progress(w *watcher) {
    s.progressIfSynced(w, []*watcher{w})
}

// progressIfSynced sends a progress to a watcher if all required watchers are synced. 
// Useful when wanting to send progress to a whole stream.
func (s *watchableStore) progressIfSynced(w *watcher, requiredSynched []*watcher) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	for _, w := range requiredSynched {
		if _, ok := s.synced.watchers[w]; !ok {
    			return false
		}
	}
	for w, _ := range watchers {
		watchers[w].send(WatchResponse{WatchID: clientv3.InvalidWatchID, Revision: s.rev()})
		return true
	}
	return true
}
```